opte/src/engine/snat.rs

```diff
@@ -295,109 +418,204 @@ impl ActionDesc for SNatDesc {
     }
 }
 
-impl Drop for SNatDesc {
+impl<T: ConcreteIpAddr> Drop for SNatDesc<T> {
     fn drop(&mut self) {
         self.pool.release(&self.priv_ip, self.nat);
     }
 }
 
+#[derive(Clone)]
+pub struct SNatIcmpEchoDesc {
+    pool: Arc<NatPool<Ipv4Addr>>,
+    nat: NatPoolEntry<Ipv4Addr>,
+    priv_ip: Ipv4Addr,
+    echo_ident: u16,
+}
+
+pub const SNAT_ICMP_ECHO_NAME: &'static str = "SNAT_ICMP_ECHO";
+
+impl ActionDesc for SNatIcmpEchoDesc {
+    fn gen_ht(&self, dir: Direction) -> HdrTransform {
+        match dir {
+            // Outbound traffic needs it's source IP and source port
+            Direction::Out => HdrTransform {
+                name: SNAT_NAME.to_string(),
+                inner_ip: Ipv4Meta::modify(Some(self.nat.ip), None, None),
+                ..Default::default()
+            },
+
+            // Inbound traffic needs its destination IP and
+            // destination port mapped back to the private values that
+            // the guest expects to see.
+            Direction::In => HdrTransform {
+                name: SNAT_NAME.to_string(),
+                inner_ip: Ipv4Meta::modify(None, Some(self.priv_ip), None),
+                ..Default::default()
+            },
+        }
+    }
+
+    // SNAT needs to generate a payload transform for ICMP traffic in
+    // order to treat the Echo Identifier as a psuedo ULP port.
+    fn gen_bt(
+        &self,
+        _dir: Direction,
+        _meta: &PacketMeta,
+        _payload_segs: &[&[u8]],
+    ) -> Result<Option<Box<dyn BodyTransform>>, GenBtError> {
+        Ok(Some(Box::new(SNatIcmpEchoBt::new(self.echo_ident, self.nat))))
+    }
+
+    fn name(&self) -> &str {
+        SNAT_ICMP_ECHO_NAME
+    }
+}
+
+impl Drop for SNatIcmpEchoDesc {
+    fn drop(&mut self) {
+        self.pool.release(&self.priv_ip.into(), self.nat);
+    }
+}
+
+/// Perform SNAT for ICMP Echo/Reply messages, treating the Identifier
+/// as a source port.
+#[derive(Clone)]
+pub struct SNatIcmpEchoBt {
+    ident: u16,
+    nat: NatPoolEntry<Ipv4Addr>,
```

Comment:
I don't think we need it now, but this would be needed for ICMPv6 echoes too, right?

Reply:
Yea I decided to just do ICMPv4 here. And I'm not sure if I would make this generic. I'd probably just make a separate type. But I imagine you'll be implementing the ICMPv6 so you can decide.

Reply:
Sounds good. I think we've been generally going towards the idea of keeping IPv4 and IPv6 separate whenever possible, to avoid all the BS you've hit around checking for impossible conditions.